apps/user-office-backend/src/factory/pdf/proposalTokenAccess.ts

```diff
@@ -0,0 +1,362 @@
+    if (questionaryId) {
+      questionarySteps = await this.questionaryDataSource.getQuestionarySteps(
+        questionaryId
+      );
+    }
+    if (isRejection(questionarySteps) || questionarySteps == null) {
+      throw new Error('Could not fetch questionary');
+    }
+
+    return questionarySteps;
+  }
+
+  private async getSampleQuestionarySteps(
+    questionaryId: number
+  ): Promise<QuestionaryStep[]> {
+    const questionarySteps =
+      await this.questionaryDataSource.getQuestionarySteps(questionaryId);
+    if (!questionarySteps) {
+      throw new Error(
+        `Questionary steps for Questionary ID '${questionaryId}' not found, or the user has insufficient rights`
+      );
+    }
+
+    return questionarySteps;
+  }
+
+  private async getSampleData() {
+    const proposalPk = this.proposalPDFData.proposal?.primaryKey;
+    this.samples = await this.sampleDataSource.getSamples({
+      filter: { proposalPk },
+    });
+
+    return (
+      await Promise.all(
+        this.samples.map(async (sample) => {
+          return collectSamplePDFData(
+            sample.id,
+            this.userWithRole as UserWithRole,
+            undefined,
+            sample,
+            await this.getQuestionary(sample.questionaryId),
+            await this.getSampleQuestionarySteps(sample.questionaryId)
+          );
+        })
+      )
+    ).map(({ sample, sampleQuestionaryFields, attachments }) => {
+      this.sampleAttachments.push(...attachments);
+
+      return { sample, sampleQuestionaryFields };
+    });
+  }
+
+  private async genericTemplatePDFData() {
+    const proposalPk = this.proposalPDFData.proposal?.primaryKey;
+    this.genericTemplates =
+      await this.genericTemplateDataSource.getGenericTemplates({
+        filter: { proposalPk },
+      });
+
+    return (
+      await Promise.all(
+        this.genericTemplates.map(async (genericTemplate) =>
+          collectGenericTemplatePDFData(
+            genericTemplate.id,
+            this.userWithRole as UserWithRole,
+            undefined,
+            genericTemplate,
+            await this.getQuestionary(genericTemplate.questionaryId),
+            await this.getSampleQuestionarySteps(genericTemplate.questionaryId)
+          )
+        )
+      )
+    ).map(
+      ({ genericTemplate, genericTemplateQuestionaryFields, attachments }) => {
+        this.genericTemplateAttachments.push(...attachments);
+
+        return { genericTemplate, genericTemplateQuestionaryFields };
+      }
+    );
+  }
+
+  private setAllTopicsInformation(out: ProposalPDFData): void {
+    // Information from each topic in proposal
+    const questionarySteps = this.proposalPDFData.questionarySteps;
+    if (questionarySteps)
+      for (const step of questionarySteps) {
+        if (!step) {
+          logger.logError('step not found', { ...questionarySteps }); // TODO: fix type of the second param in the lib (don't use Record<string, unknown>)
```

Comment:
Is this TODO something that still needs attention? If so, could it be removed and have a separate issue opened?

Reply:
Created this issue https://github.com/UserOfficeProject/user-office-project-issue-tracker/issues/761 but removing the comment without fixing the underlining problem might not be the best decision.